### PR TITLE
ext_redis.py support redis clusters --- Fixes #9538

### LINKED ISSUE #9538: Add support for Redis Cluster to ext_redis.py
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

I would like to propose a feature to address a challenge I encountered while using Dify: the current ext_redis.py does not support Redis Cluster.Due to the lack of support for Redis Cluster, my efficiency and performance when using Dify for automation are lower.I hope ext_redis.py can support Redis Cluster.Redis Cluster can promote other healthy nodes to master nodes when the current master node encounters issues, enhancing the system's robustness. It also improves the system's memory utilization and write performance, making it suitable for larger-scale and more demanding data processing scenarios.Could you please let me know if it's feasible to make ext_redis.py support Redis Cluster?

### 2. Additional context or comments

_No response_

### 3. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by liuhaoran1212**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

This submission has enabled support for starting Redis clusters in the ext_redis.py
Fixes #9538 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: api/extensions/ext_redis.py
Comment (bowenliang123) [inline api/extensions/ext_redis.py]:
```suggestion
        nodes = [
            ClusterNode(host=node.split(":")[0], port=int(node.split(":")[1])) 
            for node in dify_config.REDIS_CLUSTERS.split(",")
        ]
```
Simplified version without repeated looping.

Comment (liuhaoran1212) [reply api/extensions/ext_redis.py]: Okay, I have submitted the code according to your suggestions

File: api/configs/middleware/cache/redis_config.py
Comment (bowenliang123) [inline api/configs/middleware/cache/redis_config.py]:
```suggestion
    REDIS_USE_CLUSTERS: bool = Field(
```
`Optional` is not required as default value is always provided.

File: api/tests/unit_tests/core/test_model_manager.py
Comment (bowenliang123) [inline api/tests/unit_tests/core/test_model_manager.py]: You are pinning details the hardcoded values here. Could you change to use the constructor in order to align to the settings and test coverage?

**Comment by bowenliang123** [comment]
LGTM overall. cc @takatost @laipz8200

File: api/tests/unit_tests/core/test_model_manager.py
Comment (liuhaoran1212) [reply api/tests/unit_tests/core/test_model_manager.py]: Okay, I have submitted the code according to your suggestions，thanks

**Comment by liuhaoran1212** [comment]
@bowenliang123 @takatost @laipz8200 Can the code be merged into the main branch? Please confirm, thanks.

**Comment by laipz8200** [comment]
Please make sure the CI passes.

**Comment by liuhaoran1212** [comment]
There can not see CI is running, only display "2 workflows awaiting approval" , Why? @laipz8200 @bowenliang123

**Comment by liuhaoran1212** [comment]
CI has passed, looking forward to merging into main, thanks. @laipz8200

**Comment by laipz8200** [comment]
Please also add the new environment variables to `.env.example` files located under `api/` and `docker/`.

**Review by laipz8200 — APPROVED** [review]
Please sync with the main branch before the merge.

**Comment by boHam-wang** [comment]
How to configure CELERY_BROKER_URL using redis-cluster?

**Comment by hoshinoharu** [comment]
"FIXME: mypy error here, try to figure out how to fix it
redis_client.initialize(RedisCluster(startup_nodes=nodes, password=dify_config.REDIS_CLUSTERS_PASSWORD)) # type: ignore
I encountered a connection issue when configuring Redis in cluster mode with docker-compose. The default Redis port I set was 6379, but the logs showed the connected Redis port as 0, and the connection failed. My solution was to change the Redis cluster's mapped port to 26379:6379, then recreate the dify-api container, and the Redis connection was successful."

**Pull request merged**
